pwndbg/heap_tracking.py

```diff
@@ -0,0 +1,357 @@
+"""
+Heap Tracking
+
+This module implements runtime tracking of the heap, allowing pwndbg to detect
+heap related misbehavior coming from an inferior in real time, which lets us
+catch UAF bugs, double frees (and more), and report them to the user.
+
+# Approach
+The approach used starting with using breakpoints to hook into the following
+libc symbols: `malloc`, `free`, `calloc`, and `realloc`. Each hook has a
```

Comment:
It might be prudent to use `__libc_malloc` et al to avoid confusion with the ld.so malloc implementation.

Reply:
Turns out `__GI___libc_malloc` is re-entrant in a way I'm not sure the code supports. Is there any reason for doing that to avoid `ld.so` over checking if any of the frames in the current stack belong to the program, as the patch already does with `in_program_code_stack()`? Or perhaps we could make sure we're installing the breakpoint for the malloc we're actually interested in somehow?